File: Sharebook/Client/src/app/core/Core/Models/TwitterEntities/UserMentionEntity.ts
import {IUserMentionEntity} from "../../../Public/Models/Entities/IUserMentionEntity";

// Object storing information related with an user mention on Twitter
export class UserMentionEntity implements IUserMentionEntity {
  // [JsonProperty("id")]
  public id: number; // long?

  // [JsonProperty("id_str")]
  public idStr: string;

  // [JsonProperty("screen_name")]
  public screenName: string;

  // [JsonProperty("name")]
  public name: string;

  // [JsonProperty("indices")]
  public indices: Array<number>;

  public equals(other: IUserMentionEntity): boolean {
    if (other == null || !this.id || !other.id) {
      return false;
    }

    if (this.id !== other.id) {
      return false;
    }

    if (this.indices == null || other.indices == null) {
      return this.indices === other.indices;      // this.Equals(this.indices, other.indices);
    }

    return this.indices.containsSameObjectsAs(other.indices, true);
  }

  public toString(): string {
    return `@${(this.screenName)}`;
  }
}

File: Sharebook/Server/src/Services/Post/Sharebook.Post.Domain/AggregatesModel/PostAggregate/TwitterEntities/UserMentionEntity.cs
﻿namespace Sharebook.Post.Domain.AggregatesModel.PostAggregate.TwitterEntities
{
    using System.Collections.Generic;
    using Newtonsoft.Json;

    using Sharebook.Common.Extensions;
    using Sharebook.Storage.Domain.AggregatesModel.PostAggregate.Entities;

    /// <summary>
    /// Object storing information related with an user mention on Twitter
    /// </summary>
    public class UserMentionEntity : IUserMentionEntity
    {
        [JsonProperty("id")]
        public long? Id { get; set; }

        [JsonProperty("id_str")]
        public string IdStr { get; set; }

        [JsonProperty("screen_name")]
        public string ScreenName { get; set; }

        [JsonProperty("name")]
        public string Name { get; set; }

        [JsonProperty("indices")]
        public IList<int> Indices { get; set; }

        public bool Equals(IUserMentionEntity other)
        {
            if (other == null || !Id.HasValue || !other.Id.HasValue)
            {
                return false;
            }

            if (Id.Value != other.Id.Value)
            {
                return false;
            }

            if (Indices == null || other.Indices == null)
            {
                return Equals(Indices, other.Indices);
            }

            return Indices.ContainsSameObjectsAs(other.Indices, true);
        }

        public override string ToString()
        {
            return $"@{ScreenName}";
        }
    }
}
